src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java

```diff
@@ -0,0 +1,349 @@
+            requireNonNull(classDeclaration.getType(), "Class type must not be null");
+            final String className = requireNonNull(classDeclaration.getType().getFullyQualifiedName(),
+                    "Fully qualified name of class must not be null");
+
+            return JavaType.ShallowClass.build(className)
+                    .withKind(JavaType.FullyQualified.Kind.Record);
+        }
+
+        private static boolean isRecord(final J.ClassDeclaration classDeclaration) {
+            return J.ClassDeclaration.Kind.Type.Record.equals(classDeclaration.getKind());
+        }
+
+        private static boolean hasExplicitConstructor(final J.ClassDeclaration classDeclaration) {
+            return classDeclaration.getPrimaryConstructor() != null || classDeclaration
+                    .getBody()
+                    .getStatements()
+                    .stream()
+                    .filter(J.MethodDeclaration.class::isInstance)
+                    .map(J.MethodDeclaration.class::cast)
+                    .map(J.MethodDeclaration::getMethodType)
+                    .filter(Objects::nonNull)
+                    .anyMatch(JavaType.Method::isConstructor);
+        }
+
+        private static List<Statement> mapToConstructorArguments(
+                final List<J.VariableDeclarations> memberVariables
+        ) {
+            return memberVariables
+                    .stream()
+                    .map(it -> it
+                            .withModifiers(Collections.emptyList())
+                            .withVariables(it.getVariables())
+                    )
+                    .map(Statement.class::cast)
+                    .collect(Collectors.toList());
+        }
+
+        private J.ClassDeclaration removeValueAnnotation(final J.ClassDeclaration cd, final ExecutionContext ctx) {
+            maybeRemoveImport("lombok.Value");
+
+            return new RemoveAnnotationVisitor(
+                    new AnnotationMatcher("@lombok.Value")
+            ).visitClassDeclaration(cd, ctx);
+        }
+
+        private static List<J.VariableDeclarations> findAllClassFields(final J.ClassDeclaration cd) {
+            return new ArrayList<>(cd.getBody().getStatements())
+                    .stream()
+                    .filter(J.VariableDeclarations.class::isInstance)
+                    .map(J.VariableDeclarations.class::cast)
+                    .collect(Collectors.toList());
+        }
+
+        private static boolean hasMemberVariableAssignments(final List<J.VariableDeclarations> memberVariables) {
+            return memberVariables
+                    .stream()
+                    .map(J.VariableDeclarations::getVariables)
+                    .flatMap(List::stream)
+                    .map(J.VariableDeclarations.NamedVariable::getInitializer)
+                    .anyMatch(J.Literal.class::isInstance);
+        }
+
+        private static boolean hasOnlyLombokValueAnnotation(final J.ClassDeclaration cd) {
```

Comment:
Should we additionally check if there are any fields annotated with `@With` for instance? Annotations on any field should disqualify the class as well I suppose; not sure if that's already correctly handled, and didn't spot a corresponding case in the (extensive!) tests yet.

Reply:
Sounds like a good idea. Maybe it would be good to filter for general field annotations related to lombok for this recipe.

Reply:
I think just any field annotation would disqualify the conversion here; I'm not sure we'd correctly handle Jackson annotations either for instance, so I'd like to start out with the simplest case only: class with only fields and only a single annotation in total, such that we don't accidentally break any code with this recipe. We can always (safely) expand the scope of what we cover later; we can't easily recover trust when we break things. :)